hazelcast/src/hazelcast/client/network.cpp

```diff
@@ -183,416 +208,272 @@ namespace hazelcast {
+                                                                             client.getName(), labels_);
                     }
-                    fireConnectionAddedEvent(connection);
-                } else {
-                    if (logger.isFinestEnabled()) {
-                        logger.finest("Re-authentication succeeded for ", *connection);
+                    case security::credentials::type::secret:
+                    case security::credentials::token:
+                    {
+                        auto cr = boost::static_pointer_cast<security::token_credentials>(credential);
+                        return protocol::codec::client_authenticationcustom_encode(cluster_name, cr->get_secret(),
+                                                                                   client_uuid_, protocol::ClientTypes::CPP,
+                                                                                   serializationVersion, HAZELCAST_VERSION,
+                                                                                   client.getName(), labels_);
                     }
-                    assert(*connection == *oldConnection);
-                }
-
-                connectionsInProgress.remove(target);
-                std::ostringstream out;
-                if (connection->getRemoteEndpoint().get()) {
-                    out << *connection->getRemoteEndpoint();
-                } else {
-                    out << "null";
-                }
-                logger.info("Authenticated with server ", out.str(), ", server version:",
-                            connection->getConnectedServerVersionString(), " Local address: ",
-                            (connection->getLocalSocketAddress().get() != NULL
-                             ? connection->getLocalSocketAddress()->toString() : "null"));
-
-                /* check if connection is closed by remote before authentication complete, if that is the case
-                we need to remove it back from active connections.
-                Race description from https://github.com/hazelcast/hazelcast/pull/8832.(A little bit changed)
-                - open a connection client -> member
-                - send auth message
-                - receive auth reply -> reply processing is offloaded to an executor. Did not start to run yet.
-                - member closes the connection -> the connection is trying to removed from map
-                                                                     but it was not there to begin with
-                - the executor start processing the auth reply -> it put the connection to the connection map.
-                - we end up with a closed connection in activeConnections map */
-                if (!connection->isAlive()) {
-                    removeFromActiveConnections(connection);
                 }
+                return protocol::ClientMessage();
```

Comment:
This path looks really wierd. Can we put an assert or throw exception to emphasize that this path is not possible.

Reply:
It was there to make the compiler happy but surely we can put an assert. putting `assert(0);`